numba/core/config.py

```diff
@@ -322,8 +322,21 @@ def avx_default():
         NUMBA_DEFAULT_NUM_THREADS = max(1, multiprocessing.cpu_count())
 
         # Numba thread pool size (defaults to number of CPUs on the system).
-        NUMBA_NUM_THREADS = _readenv("NUMBA_NUM_THREADS", int,
-                                     NUMBA_DEFAULT_NUM_THREADS)
+        _NUMBA_NUM_THREADS = _readenv("NUMBA_NUM_THREADS", int,
+                                      NUMBA_DEFAULT_NUM_THREADS)
+        if ('NUMBA_NUM_THREADS' in globals()
+                and globals()['NUMBA_NUM_THREADS'] != _NUMBA_NUM_THREADS):
+
+            from numba.np.ufunc import parallel
+            if parallel._is_initialized:
+                raise RuntimeError("Cannot set NUMBA_NUM_THREADS to a "
+                                   "different value once the threads have been "
+                                   "launched (%s != %s)" %
+                                   (_NUMBA_NUM_THREADS,
+                                    globals()['NUMBA_NUM_THREADS']))
+
+        NUMBA_NUM_THREADS = _NUMBA_NUM_THREADS
+        del _NUMBA_NUM_THREADS
```

Comment:
This logic is sufficiently complicated I think it should have test(s) added.

Reply:
Technically the test_parallel_env_var test tests this. It's not ideal though because the actual thing it tests depends on whether threads have been launched or not before it is run. 

This logic should probably be factored out. I didn't do it here because it would require refactoring this whole file and I wanted to be sure this is a good thing to do first.

Reply:
Something along the lines of

```py
def _read_config_from_env(name, typ, default_value, allow_reload=True):
    if not allow_reload and name in globals() and globals()[name] != default_value:
        raise RuntimeError
    globals()[name] = _readenv('NUMBA_' + name, typ, default_value)
```

(there's also the annoyance I mentioned earlier that `NUMBA_NUM_THREADS` and `NUMBA_DEFAULT_NUM_THREADS` don't have the "NUMBA_" stripped off them in the config)

Reply:
I think that should work (but perhaps have the `RuntimeError` message as an arg too?) . Given the `config.NUMBA_` inconsistency annoyance, seems like it'd be convenient as a reason to fix that too? :)

Reply:
I'm working on this but it it looking like something that would be better as a separate PR. If it is OK with you, I'd prefer to do that.

Reply:
I pushed my work so far to the `config-cleanup` branch on the Quansight fork. It's based off this PR, so I'll open a new PR once this PR is merged.